File: src/main/java/me/senseiwells/arucas/utils/ReflectionUtils.java
package me.senseiwells.arucas.utils;

import me.senseiwells.arucas.api.ISyntax;
import me.senseiwells.arucas.extensions.util.JavaValue;
import me.senseiwells.arucas.throwables.CodeError;
import me.senseiwells.arucas.throwables.RuntimeError;
import me.senseiwells.arucas.values.Value;

import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;
import java.lang.reflect.*;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class ReflectionUtils {
	private static final Map<MethodId, Method> methodCache = new HashMap<>();

	public static Value callMethodFromNameAndArgs(Class<?> callingClass, Object callingObject, String methodName, List<Value> arguments, ISyntax syntaxPosition, Context context) throws CodeError {
		MethodId methodId = new MethodId(callingClass, methodName, arguments.size());
		Object[] parameters = new Object[arguments.size()];

		Method wantedMethod = methodCache.get(methodId);
		if (wantedMethod == null) {
			Class<?>[] types = new Class[arguments.size()];

			for (int i = 0; i < arguments.size(); i++) {
				parameters[i] = arguments.get(i).asJavaValue();
				if (parameters[i] != null) {
					types[i] = parameters[i].getClass();
					continue;
				}
				types[i] = Object.class;
			}

			// This won't work if the parameter has primitive parameters since it converted to Wrappers
			// This will be fine since #getMethodSlow will still resolve the correct method...
			wantedMethod = ExceptionUtils.catchAsNull(() -> callingClass.getMethod(methodName, types));
			if (wantedMethod == null) {
				wantedMethod = ExceptionUtils.catchAsNull(() -> callingClass.getMethod(methodName, Object[].class));
				if (wantedMethod == null) {
					wantedMethod = getMethodSlow(callingClass, callingObject, methodName, parameters.length);
					if (wantedMethod == null) {
						throw new RuntimeError(
							"No such method '%s' with %d parameters exists for '%s'".formatted(methodName, parameters.length, callingClass.getSimpleName()),
							syntaxPosition, context
						);
					}
				}
			}
			methodCache.put(methodId, wantedMethod);
		}
		else {
			for (int i = 0; i < arguments.size(); i++) {
				parameters[i] = arguments.get(i).asJavaValue();
			}
		}

		return invokeMethod(context, syntaxPosition, wantedMethod, callingObject, parameters);
	}

	public static Value callMethodFromJavaValue(JavaValue javaValue, String methodName, List<Value> arguments, ISyntax syntaxPosition, Context context) throws CodeError {
		Object callingObject = javaValue.asJavaValue();
		Class<?> callingClass = callingObject.getClass();

		Method wantedMethod = getMethod(callingClass, callingObject, methodName, arguments.size());
		if (wantedMethod == null) {
			return null;
		}

		// This method doesn't visit arguments unless it knows that there is a valid method
		// that can be called, we don't want arguments to be visited for no reason
		Object[] objects = new Object[arguments.size()];
		for (int i = 0; i < arguments.size(); i++) {
			objects[i] = arguments.get(i).asJavaValue();
		}

		return invokeMethod(context, syntaxPosition, wantedMethod, callingObject, objects);
	}

	public static MethodHandle getMethodHandle(Method method) {
		return ExceptionUtils.catchAsNull(() -> {
			MethodHandles.Lookup lookup = MethodHandles.publicLookup();
			MethodType methodType = MethodType.methodType(method.getReturnType(), method.getParameterTypes());

			return Modifier.isStatic(method.getModifiers()) ?
				lookup.findStatic(method.getDeclaringClass(), method.getName(), methodType) :
				lookup.findVirtual(method.getDeclaringClass(), method.getName(), methodType);
		});
	}

	public static Value constructFromArgs(Class<?> constructingClass, List<Value> arguments, ISyntax syntaxPosition, Context context) throws CodeError {
		Object[] objects = new Object[arguments.size()];
		Class<?>[] parameters = new Class[arguments.size()];

		for (int i = 0; i < arguments.size(); i++) {
			objects[i] = arguments.get(i).asJavaValue();
			if (objects[i] != null) {
				parameters[i] = objects[i].getClass();
				continue;
			}
			parameters[i] = Object.class;
		}

		Constructor<?> constructor = ExceptionUtils.catchAsNull(() -> constructingClass.getConstructor(parameters));
		if (constructor == null) {
			constructor = ExceptionUtils.catchAsNull(() -> constructingClass.getConstructor(Object[].class));
			if (constructor == null) {
				constructor = getConstructorSlow(constructingClass, objects.length);
				if (constructor == null) {
					throw new RuntimeError(
						"No such constructor with %d parameters exists for '%s'".formatted(objects.length, constructingClass.getSimpleName()),
						syntaxPosition, context
					);
				}
			}
		}

		return invokeConstructor(context, syntaxPosition, constructor, objects);
	}

	public static Value getFieldFromName(Class<?> callingClass, Object callingObject, String fieldName, ISyntax syntaxPosition, Context context) throws RuntimeError {
		Field field = ExceptionUtils.catchAsNull(() -> callingClass.getField(fieldName));
		if (field != null) {
			return getField(context, syntaxPosition, field, callingObject);
		}
		throw new RuntimeError(
			"No such field '%s'".formatted(fieldName),
			syntaxPosition, context
		);
	}

	public static Value getFieldFromJavaValue(JavaValue javaValue, String fieldName, ISyntax syntaxPosition, Context context) {
		Object callingObject = javaValue.asJavaValue();
		Class<?> callingClass = callingObject.getClass();
		Field field = ExceptionUtils.catchAsNull(() -> callingClass.getField(fieldName));
		if (field != null) {
			return ExceptionUtils.catchAsNull(() -> getField(context, syntaxPosition, field, callingObject));
		}
		return null;
	}

	public static void setFieldFromName(Class<?> callingClass, Object callingObject, Object newValue, String fieldName, ISyntax syntaxPosition, Context context) throws RuntimeError {
		Field field = ExceptionUtils.catchAsNull(() -> callingClass.getField(fieldName));
		if (field != null) {
			setField(context, syntaxPosition, field, callingObject, newValue);
			return;
		}
		throw new RuntimeError(
			"No such field '%s'".formatted(fieldName),
			syntaxPosition, context
		);
	}

	public static boolean setFieldFromJavaValue(JavaValue javaValue, Value newValue, String fieldName, ISyntax syntaxPosition, Context context) {
		Object callingObject = javaValue.asJavaValue();
		Class<?> callingClass = callingObject.getClass();
		Field field = ExceptionUtils.catchAsNull(() -> callingClass.getField(fieldName));
		if (field != null) {
			return ExceptionUtils.runSafe(() -> setField(context, syntaxPosition, field, callingObject, newValue.asJavaValue()));
		}
		return false;
	}

	public static Method getMethod(Class<?> callingClass, Object callingObject, String methodName, int arguments) {
		MethodId methodId = new MethodId(callingClass, methodName, arguments);
		Method cachedMethod = methodCache.get(methodId);
		if (cachedMethod == null) {
			Method method = getMethodSlow(callingClass, callingObject, methodName, arguments);
			if (method != null) {
				methodCache.put(methodId, method);
			}
			return method;
		}
		return cachedMethod;
	}

	private static Method getMethodSlow(Class<?> callingClass, Object callingObject, String methodName, int arguments) {
		boolean isStatic = callingObject == null;
		for (Method method : callingClass.getMethods()) {
			boolean matchingStatic = Modifier.isStatic(method.getModifiers()) == isStatic;
			if (matchingStatic && (arguments < 0 || method.getParameterCount() == arguments) && method.getName().equals(methodName)) {
				Method accessible = getAccessibleMethod(callingClass, callingObject, method);
				return Objects.requireNonNullElse(accessible, method);
			}
		}
		return null;
	}

	private static Method getAccessibleMethod(Class<?> callingClass, Object callingObject, Method method) {
		// The direct method may be inaccessible (for example it's a member of an inaccessible inner class)
		// but accessible via an interface or a superclass. Notably this happens for synthetic lambda classes.
		// This is a workaround which looks for an accessible method in a superclass or interface.

		if (method.canAccess(callingObject)) {
			return method;
		}

		try {
			method = callingClass.getMethod(method.getName(), method.getParameterTypes());
			if (method.canAccess(callingObject)) {
				return method;
			}
		} catch (NoSuchMethodException e) {
			return null;
		}

		if (!callingClass.isInterface()) {
			Class<?> superClass = callingClass.getSuperclass();
			if (superClass != null) {
				Method accessible = getAccessibleMethod(superClass, callingObject, method);
				if (accessible != null) {
					return accessible;
				}
			}
		}

		for (Class<?> itf : callingClass.getInterfaces()) {
			Method accessible = getAccessibleMethod(itf, callingObject, method);
			if (accessible != null) {
				return accessible;
			}
		}

		return null;
	}

	private static Constructor<?> getConstructorSlow(Class<?> callingClass, int arguments) {
		for (Constructor<?> constructor : callingClass.getConstructors()) {
			if (constructor.getParameterCount() == arguments) {
				return constructor;
			}
		}
		return null;
	}

	private static Value invokeMethod(Context context, ISyntax syntaxPosition, Method method, Object object, Object... arguments) throws CodeError {
		try {
			return JavaValue.of(method.invoke(object, arguments));
		}
		catch (InvocationTargetException e) {
			Throwable throwable = e.getCause();
			if (throwable instanceof CodeError codeError) {
				throw codeError;
			}
			throw new RuntimeError(
				"Failed to call '%s': %s".formatted(method.getName(), throwable),
				syntaxPosition, context
			);
		}
		catch (IllegalAccessException e) {
			throw new RuntimeError(
				"No access to the method '%s'".formatted(method.getName()),
				syntaxPosition, context
			);
		}
		catch (IllegalArgumentException e) {
			throw new RuntimeError(
				"Illegal Arguments for the method '%s': %s".formatted(method.getName(), e),
				syntaxPosition, context
			);
		}
	}

	private static Value invokeConstructor(Context context, ISyntax syntaxPosition, Constructor<?> constructor, Object... arguments) throws CodeError {
		try {
			return JavaValue.of(constructor.newInstance(arguments));
		}
		catch (InvocationTargetException e) {
			Throwable throwable = e.getCause();
			if (throwable instanceof CodeError codeError) {
				throw codeError;
			}
			throw new RuntimeError(
				"Failed to call '%s': %s".formatted(constructor.getName(), throwable),
				syntaxPosition, context
			);
		}
		catch (IllegalAccessException e) {
			throw new RuntimeError(
				"No access to the constructor '%s'".formatted(constructor.getName()),
				syntaxPosition, context
			);
		}
		catch (IllegalArgumentException e) {
			throw new RuntimeError(
				"Illegal Arguments for the constructor '%s': %s".formatted(constructor.getName(), e),
				syntaxPosition, context
			);
		}
		catch (InstantiationException e) {
			throw new RuntimeError(
				"Failed to construct '%s': %s".formatted(constructor.getName(), e),
				syntaxPosition, context
			);
		}
	}

	private static Value getField(Context context, ISyntax syntaxPosition, Field field, Object object) throws RuntimeError {
		try {
			// Cannot do catchAsNull since actual return value may be null
			return JavaValue.of(field.get(object));
		}
		catch (IllegalAccessException e) {
			throw new RuntimeError(
				"No access to the field '%s'".formatted(field.getName()),
				syntaxPosition, context
			);
		}
	}

	private static void setField(Context context, ISyntax syntaxPosition, Field field, Object object, Object newValue) throws RuntimeError {
		Throwable throwable = ExceptionUtils.returnThrowable(() -> field.set(object, newValue));
		if (throwable != null) {
			throw new RuntimeError(
				"Failed to set field '%s' with value: %s".formatted(field.getName(), throwable),
				syntaxPosition, context
			);
		}
	}

	private static final class MethodId {
		private final Class<?> clazz;
		private final String name;
		private final int parameters;
		private final int hash;

		private MethodId(Class<?> clazz, String name, int parameters) {
			this.clazz = clazz;
			this.name = name;
			this.parameters = parameters;
			this.hash = (clazz.hashCode() - this.name.hashCode() * this.parameters) * 31;
		}

		@Override
		public int hashCode() {
			return this.hash;
		}

		@Override
		public boolean equals(Object obj) {
			if (obj == this) {
				return true;
			}
			if (obj instanceof MethodId id) {
				return this.clazz == id.clazz && this.parameters == id.parameters && this.name.equals(id.name);
			}
			return false;
		}
	}
}

File: src/main/kotlin/me/senseiwells/arucas/utils/ReflectionUtils.kt
package me.senseiwells.arucas.utils

import me.senseiwells.arucas.api.ArucasObfuscator
import me.senseiwells.arucas.builtin.StringDef
import me.senseiwells.arucas.classes.instance.ClassInstance
import me.senseiwells.arucas.core.Interpreter
import me.senseiwells.arucas.exceptions.RuntimeError
import me.senseiwells.arucas.exceptions.runtimeError
import net.bytebuddy.implementation.bind.annotation.AllArguments
import net.bytebuddy.implementation.bind.annotation.Origin
import net.bytebuddy.implementation.bind.annotation.RuntimeType
import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodHandles
import java.lang.invoke.VarHandle
import java.lang.invoke.WrongMethodTypeException
import java.lang.reflect.*

object ReflectionUtils {
    private val methodCache = HashMap<MethodId, MethodWithHandle>()
    private val constructorCache = HashMap<ConstructorId, MethodWithHandle>()
    private val fieldCache = HashMap<FieldId, FieldWithHandle>()

    fun callMethod(callingClass: Class<*>, callingObject: Any?, name: String, arguments: List<ClassInstance>, obfuscator: ArucasObfuscator): Any? {
        val (args, types) = this.argumentsToJava(arguments)
        val id = MethodId(callingClass, name, types)
        var methodWithHandle = this.methodCache[id]
        if (methodWithHandle == null) {
            val method = this.getMethod(callingClass, callingObject, obfuscator.obfuscateMethodName(callingClass, name), types)
            if (method == null) {
                val typeNames = types.joinToString(prefix = "(", postfix = ")") { obfuscator.deobfuscateClass(it) }
                runtimeError("No such method '$name' with type parameters $typeNames exists for Java type '${obfuscator.deobfuscateClass(callingClass)}'")
            }
            methodWithHandle = this.methodWithHandle(method)
            this.methodCache[id] = methodWithHandle
        }
        if (callingObject != null) {
            args.add(0, callingObject)
        }
        return this.invokeHandle(methodWithHandle, args)
    }

    fun callConstructor(clazz: Class<*>, arguments: List<ClassInstance>, obfuscator: ArucasObfuscator): Any? {
        val (args, types) = this.argumentsToJava(arguments)
        val id = ConstructorId(clazz, types)
        var methodHandle = this.constructorCache[id]
        if (methodHandle == null) {
            val constructor = this.getConstructor(clazz, types)
            if (constructor == null) {
                val typeNames = types.joinToString(prefix = "(", postfix = ")") { obfuscator.deobfuscateClass(it) }
                runtimeError("No such constructor with parameters $typeNames exists for ${obfuscator.deobfuscateClass(clazz)}")
            }
            methodHandle = this.methodWithHandle(constructor)
            this.constructorCache[id] = methodHandle
        }
        return this.invokeHandle(methodHandle, args)
    }

    fun getField(callingClass: Class<*>, callingObject: Any?, name: String, obfuscator: ArucasObfuscator): Any? {
        val handle = this.getVarHandle(callingClass, name, obfuscator)
        return this.getVarHandle(handle, callingObject)
    }

    fun setField(callingClass: Class<*>, callingObject: Any?, setObject: ClassInstance, name: String, obfuscator: ArucasObfuscator) {
        val handle = this.getVarHandle(callingClass, name, obfuscator)
        this.setVarHandle(handle, callingObject, setObject.asJava())
    }

    fun getClass(name: String, obfuscator: ArucasObfuscator): Class<*> {
        val obfuscatedName = obfuscator.obfuscateClassName(name)
        try {
            return Class.forName(obfuscatedName)
        } catch (e: Exception) {
            runtimeError("Failed to get Java class", e)
        }
    }

    fun functionToInterceptor(interpreter: Interpreter, function: ClassInstance): Any {
        return Interceptor(interpreter, function)
    }

    private fun getVarHandle(clazz: Class<*>, name: String, obfuscator: ArucasObfuscator): FieldWithHandle {
        val id = FieldId(clazz, name)
        var varHandle = this.fieldCache[id]
        if (varHandle == null) {
            val field = this.getField(clazz, obfuscator.obfuscateFieldName(clazz, name))
                ?: runtimeError("No such field with name '${name}' exists for ${obfuscator.deobfuscateClass(clazz)}")
            varHandle = this.getVarHandle(field)
            this.fieldCache[id] = varHandle
        }
        return varHandle
    }

    private fun argumentsToJava(arguments: List<ClassInstance>): Pair<ArrayList<Any?>, Array<Class<*>>> {
        val args = ArrayList<Any?>(arguments.size)
        val types = Array<Class<*>>(arguments.size) { i ->
            val argument = arguments[i].asJava()
            args.add(argument)
            if (argument != null) {
                this.wrapClass(argument::class.java)
            } else {
                Void.TYPE
            }
        }
        return args to types
    }

    private fun getMethod(callingClass: Class<*>, callingObject: Any?, name: String, types: Array<Class<*>>): Method? {
        val isStatic = callingObject == null
        for (method in callingClass.methods) {
            val matchStatic = Modifier.isStatic(method.modifiers) == isStatic
            if (matchStatic && method.name == name && this.doParametersMatch(method, types)) {
                val accessible = this.getAccessibleMethod(callingClass, callingObject, method)
                return accessible ?: method
            }
        }
        return null
    }

    private fun getConstructor(clazz: Class<*>, types: Array<Class<*>>): Constructor<*>? {
        for (constructor in clazz.constructors) {
            if (this.doParametersMatch(constructor, types)) {
                return constructor
            }
        }
        return null
    }

    private fun getField(clazz: Class<*>, name: String): Field? {
        for (field in clazz.fields) {
            if (field.name == name) {
                return field
            }
        }
        return null
    }

    private fun doParametersMatch(executable: Executable, given: Array<Class<*>>): Boolean {
        val required = executable.parameterTypes

        if (executable.isVarArgs) {
            // Technically VarArgs can be empty
            if (required.size - 1 > given.size) {
                return false
            }
        } else if (required.size != given.size) {
            return false
        }

        for (i in required.indices) {
            if (executable.isVarArgs && i == required.lastIndex) {
                val component = required[i].componentType
                for (gi in i..given.lastIndex) {
                    if (!areTypesEqual(given[gi], component)) {
                        return false
                    }
                }
                return true
            }
            if (i > given.lastIndex) {
                return false
            }

            if (!areTypesEqual(given[i], required[i])) {
                return false
            }
        }

        return true
    }

    private fun areTypesEqual(given: Class<*>, required: Class<*>): Boolean {
        if (required.isPrimitive && given == Void.TYPE) {
            return false
        }
        val requiredWrapped = this.wrapClass(required)
        // Something that we allow is casting from any number type to any number
        if (Number::class.java.isAssignableFrom(requiredWrapped) && Number::class.java.isAssignableFrom(given)) {
            return true
        }
        if (given != Void.TYPE && !requiredWrapped.isAssignableFrom(given)) {
            return false
        }
        return true
    }

    private fun getAccessibleMethod(callingClass: Class<*>, callingObject: Any?, method: Method): Method? {
        // The direct method may be inaccessible (for example it's a member of an inaccessible inner class)
        // but accessible via an interface or a superclass. Notably this happens for synthetic lambda classes.
        // This is a workaround which looks for an accessible method in a superclass or interface.

        if (method.canAccess(callingObject)) {
            return method
        }

        var mutMethod = method
        try {
            mutMethod = callingClass.getMethod(mutMethod.name, *mutMethod.parameterTypes)
            if (mutMethod.canAccess(callingObject)) {
                return mutMethod
            }
        } catch (_: NoSuchMethodException) {
            return null
        }

        if (!callingClass.isInterface) {
            val superClass = callingClass.superclass
            if (superClass != null) {
                val accessible = this.getAccessibleMethod(superClass, callingObject, mutMethod)
                if (accessible != null) {
                    return accessible
                }
            }
        }

        for (itf in callingClass.interfaces) {
            val accessible = this.getAccessibleMethod(itf, callingObject, mutMethod)
            if (accessible != null) {
                return accessible
            }
        }

        return null
    }

    private fun methodWithHandle(method: Method): MethodWithHandle {
        return MethodWithHandle(method, MethodHandles.publicLookup().unreflect(method))
    }

    private fun methodWithHandle(constructor: Constructor<*>): MethodWithHandle {
        return MethodWithHandle(constructor, MethodHandles.publicLookup().unreflectConstructor(constructor))
    }

    private fun getVarHandle(field: Field): FieldWithHandle {
        return FieldWithHandle(field, MethodHandles.publicLookup().unreflectVarHandle(field))
    }

    private fun invokeHandle(methodHandle: MethodWithHandle, args: List<Any?>): Any? {
        try {
            return methodHandle.invoke(args)
        } catch (e: RuntimeError) {
            throw e
        } catch (e: ClassCastException) {
            runtimeError("Incorrect parameter type was found", e)
        } catch (e: WrongMethodTypeException) {
            runtimeError("Method was invoked incorrectly", e)
        } catch (e: Exception) {
            runtimeError("An unexpected error was thrown during java method call", e)
        }
    }

    private fun setVarHandle(varHandle: FieldWithHandle, calling: Any?, value: Any?) {
        try {
            varHandle.set(calling, value)
        } catch (e: RuntimeError) {
            throw e
        } catch (e: ClassCastException) {
            runtimeError("Incorrect object for field assignment", e)
        } catch (e: WrongMethodTypeException) {
            val error = if (calling == null) "statically but expected virtual" else "virtually but expected static"
            runtimeError("Field was set $error assignment", e)
        } catch (e: UnsupportedOperationException) {
            runtimeError("Field cannot be reassigned", e)
        } catch (e: Exception) {
            runtimeError("An unexpected error was thrown during java field assignment", e)
        }
    }

    private fun getVarHandle(varHandle: FieldWithHandle, calling: Any?): Any? {
        try {
            return varHandle.get(calling)
        } catch (e: RuntimeError) {
            throw e
        } catch (e: ClassCastException) {
            runtimeError("Incorrect object for field access", e)
        } catch (e: WrongMethodTypeException) {
            val error = if (calling == null) "statically but expected virtual" else "virtually but expected static"
            runtimeError("Field was accessed $error access", e)
        } catch (e: Exception) {
            runtimeError("An unexpected error was thrown during java field access", e)
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> wrapClass(type: Class<T>): Class<T> {
        return when (type) {
            Boolean::class.java -> java.lang.Boolean::class.java
            Byte::class.java -> java.lang.Byte::class.java
            Char::class.java -> java.lang.Character::class.java
            Double::class.java -> java.lang.Double::class.java
            Float::class.java -> java.lang.Float::class.java
            Int::class.java -> java.lang.Integer::class.java
            Long::class.java -> java.lang.Long::class.java
            Short::class.java -> java.lang.Short::class.java
            Void.TYPE -> Void::class.java
            else -> type
        } as Class<T>
    }

    private fun castToNumber(any: Any?, clazz: Class<*>): Any? {
        if (any !is Number) {
            return any
        }
        return when (clazz) {
            java.lang.Byte::class.java -> any.toByte()
            java.lang.Short::class.java -> any.toShort()
            java.lang.Integer::class.java -> any.toInt()
            java.lang.Long::class.java -> any.toLong()
            java.lang.Float::class.java -> any.toFloat()
            java.lang.Double::class.java -> any.toDouble()
            else -> any // ???
        }
    }

    private data class MethodId(val clazz: Class<*>, val name: String, val types: Array<Class<*>>) {
        val hash: Int

        init {
            var hash = this.clazz.hashCode()
            hash = 31 * hash + this.name.hashCode()
            hash = 31 * hash + this.types.contentHashCode()
            this.hash = hash
        }

        override fun equals(other: Any?): Boolean {
            if (other !is MethodId || this.clazz != other.clazz || this.name != other.name) {
                return false
            }
            return this.types.contentEquals(other.types)
        }

        override fun hashCode() = this.hash
    }

    private data class ConstructorId(val clazz: Class<*>, val types: Array<Class<*>>) {
        val hash = 31 * this.clazz.hashCode() + this.types.contentHashCode()

        override fun equals(other: Any?): Boolean {
            if (other !is ConstructorId || this.clazz != other.clazz) {
                return false
            }
            return this.types.contentEquals(other.types)
        }

        override fun hashCode() = this.hash
    }

    private data class FieldId(val clazz: Class<*>, val name: String) {
        val hash = 31 * this.clazz.hashCode() + this.name.hashCode()

        override fun equals(other: Any?): Boolean {
            return other is FieldId && this.clazz == other.clazz && this.name == other.name
        }

        override fun hashCode(): Int {
            return this.hash
        }
    }

    private data class MethodWithHandle(val method: Executable, val handle: MethodHandle) {
        private val casts by lazy {
            val casts = HashMap<Int, (Any?) -> Any?>()
            this.method.parameterTypes.forEachIndexed { index, clazz ->
                if (this.method.isVarArgs && index == this.method.parameterCount - 1) {
                    val wrapped = wrapClass(clazz.componentType)
                    if (Number::class.java.isAssignableFrom(wrapped)) {
                        casts[index] = { castToNumber(it, wrapped) }
                    }
                    return@forEachIndexed
                }
                val wrapped = wrapClass(clazz)
                if (Number::class.java.isAssignableFrom(wrapped)) {
                    casts[index] = { castToNumber(it, wrapped) }
                }
            }
            casts
        }

        fun invoke(arguments: List<Any?>): Any? {
            val mapped = arguments.mapIndexed { index, any ->
                this.casts[index.coerceAtMost(this.method.parameterCount - 1)]?.invoke(any) ?: any
            }
            return handle.invokeWithArguments(mapped)
        }
    }

    private data class FieldWithHandle(val field: Field, val handle: VarHandle) {
        private val fieldType = wrapClass(field.type)
        private val shouldTryToCast = Number::class.java.isAssignableFrom(fieldType)

        fun set(calling: Any?, value: Any?) {
            val castedValue = if (this.shouldTryToCast) castToNumber(value, this.fieldType) else value
            if (calling == null) this.handle.set(castedValue) else this.handle.set(calling, castedValue)
        }

        fun get(calling: Any?): Any? {
            return if (calling == null) this.handle.get() else this.handle.get(calling)
        }
    }

    class Interceptor(interpreter: Interpreter, val function: ClassInstance) {
        val interpreter = interpreter.branch()

        @RuntimeType
        fun intercept(@AllArguments allArguments: Array<Any?>, @Origin method: Method) {
            val branch = this.interpreter.branch()
            val arguments = ArrayList<ClassInstance>()
            arguments.add(branch.create(StringDef::class, method.name))
            arguments.add(branch.convertValue(allArguments))
            branch.call(this.function, arguments)
        }
    }
}